Sources/Internal/Transformations/Spline.h

```diff
@@ -41,29 +42,163 @@ namespace DataAnalysis { namespace Transformations {
 			}
 		}
 
+		/*
+			Interpolates the spline at given point
+			Should be done only between given control points
+		*/
 		virtual inline void Apply( __in const BaseType &in, __out BaseType &out ) const {
-			out = in;
+			Segment *pSegment = GetSegment( in );
+
+			BaseType xScale = pSegment->pCpEnd->x - pSegment->pCpStart->x;
+			BaseType inScaled = ( in - pSegment->pCpStart->x ) / xScale;
+			BaseType S[4] = { inScaled*inScaled*inScaled, inScaled*inScaled, inScaled, 1 };
+			BaseType C[4] = {   pSegment->pCpStart->fx,
+								pSegment->pCpEnd->fx,
+								pSegment->pCpStart->t * xScale,
+								pSegment->pCpEnd->t * xScale };
+
+			// P = S * h * C
+			BaseType tmp[4] = { 0, 0, 0, 0 };
+			Vect4MulMat44( S, m_basisFunctionsT, tmp );
+			Vect4DotProduct( tmp, C, &out );
 		}
 
 	protected:
-		size_t mSplinePointCount;
-		Buffer< CubicSplineParamPair<BaseType> > mSplinePoints;
+
+		struct ControlPoint {
+			BaseType x;
+			BaseType t; // tangent ( differentiation )
+			BaseType fx;
+		};
+
+		struct Segment {
+			ControlPoint *pCpStart;
+			ControlPoint *pCpEnd;
+		};
+
+
+		size_t m_segmentCount;
+		size_t m_cpCount;
+		Buffer< ControlPoint > m_controlPoints;
+		Buffer< Segment > m_segments;
+		
+		const BaseType m_basisFunctionsT[16] = {  2, -2,  1,  1,
+												 -3,  3, -2, -1,
+												  0,  0,  1,  0,
+												  1,  0,  0,  0 };
 
 	protected:
 
-		inline void InitializeVars( __in const size_t count ) {
-			mSplinePointCount = count;
-			mSplinePoints.Allocate( count );
+		/*
+		Allocates buffer for holding spline segment data
+		There are ptCount + 1 segments, because we do include segments [-inf, CP1_X] and [CPn_X, inf]
```

Comment:
because with N points, you get (N-1) inner segmets and when you add 2 outer segments, you get total of "(N-1) + 2 = (N + 1)" segments